File: modules/algebra/test/test_geometric_hash.py
import math
import IMP.algebra
import IMP.test

# square Euclidean distance
def sq_dist(p1, p2):
    return (p1[0] - p2[0])**2 + (p1[1] - p2[1])**2


# find the points within the distance of radius from the center - the slow way
def slow_nearest(points, center, radius):
    radius *= radius
    r = list()
    for p in points:
        if sq_dist(p, center) <= radius:
            r.append(p)
    return r


# find the points within the distance of radius from the center - the "fast" way
def fast_nearest(points, center, radius):
    H = IMP.algebra.GeometricHash2()
    for i, p in enumerate(points):
        H.add(p, i)
    v = H.neighbors(H.EUCLIDEAN, center, radius)
    return [points[x] for x in v]


# generate some points
def gen_points(N):
    points = list()
    for r in xrange(N):
        for i in xrange(N):
            angle = 2.0*i*math.pi/N
            p = [0.1*r*math.cos(angle), 0.1*r*math.sin(angle)]
            points.append(p)
    return points


# compare the points generated using the slow and fast algorithm
def compare(slow, fast):
    if len(slow) != len(fast):
        return False
    slow.sort(key = lambda x : x[1])
    slow.sort(key = lambda x : x[0])
    fast.sort(key = lambda x : x[1])
    fast.sort(key = lambda x : x[0])
    for s, f in zip(slow, fast):
        if s[0] != f[0] or s[1] != f[1]:
            return False
    return True



class GeometricHashTest(IMP.test.TestCase):
    def test_neighbors(self):
        points = gen_points(200)
        cnt = [1, 1]
        rd = 0.45
        slow = slow_nearest(points, cnt, rd)
        fast = fast_nearest(points, cnt, rd)
        status = compare(slow, fast)
        self.assertEqual(status, True)


if __name__ == '__main__':
    IMP.test.main()

File: modules/multifit/test/test_geometric_hash.cpp
/*   Test the GeometricHash class by comparing its results to those obtained
 *   by simple distance calculation.
 *
 *   Copyright 2007-2013 IMP Inventors. All rights reserved.
 */

#include <IMP/algebra/constants.h>
#include <IMP/multifit/GeometricHash.h>

typedef IMP::multifit::GeometricHash<int, 2> Hash2;
typedef std::vector<Hash2::Point> Points;
typedef std::vector<int> Indices;

namespace {
// square Euclidean distance
double sq_dist(const Hash2::Point &p1, const Hash2::Point &p2)
{
  double d0 = p1[0] - p2[0];
  double d1 = p1[1] - p2[1];
  return d0*d0 + d1*d1;
}

// find the indices of points within the distance of radius from the
// center by simple distance calculation
Indices dist_nearest(const Points &points, const Hash2::Point &center,
                     double radius)
{
  radius *= radius;
  Indices inds;
  for (unsigned i = 0; i < points.size(); ++i) {
    if (sq_dist(points[i], center) <= radius) {
      inds.push_back(i);
    }
  }
  return inds;
}


// find the indices of points within the distance of radius from the
// center using the hash
Indices hash_nearest(const Points &points, const Hash2::Point &center,
                     double radius)
{
  Hash2 H;
  for (unsigned i = 0; i < points.size(); ++i) {
    H.add(points[i], i);
  }
  Hash2::HashResult r = H.neighbors(H.EUCLIDEAN, center, radius);
  Indices ret;
  for (Hash2::HashResult::const_iterator it = r.begin(); it != r.end(); ++it) {
    ret.push_back((*it)->second);
  }
  return ret;
}

// generate some points
Points gen_points(int N)
{
  Points points;
  for (int r = 0; r < N; ++r) {
    for (int i = 0; i < N; ++i) {
      double angle = 2.0*i*IMP::algebra::PI/static_cast<float>(N);
      points.push_back(Hash2::Point(0.1*r*std::cos(angle),
                                    0.1*r*std::sin(angle)));
    }
  }
  return points;
}
}
int main()
{
  Points points = gen_points(200);
  Hash2::Point cnt(1.,1.);
  double rd = 0.45;
  Indices dist = dist_nearest(points, cnt, rd);
  Indices hash = hash_nearest(points, cnt, rd);

  std::sort(dist.begin(), dist.end());
  std::sort(hash.begin(), hash.end());

  if (dist.size() != hash.size()) {
    IMP_THROW("lists are different sizes", ValueException);
  }

  if (!std::equal(dist.begin(), dist.end(), hash.begin())) {
    IMP_THROW("lists have differing elements", ValueException);
  }

  return 0;
}
